### PR TITLE
Prebuild Events

### PR DESCRIPTION
**Pull request by AlexTugarev**
This PR includes

 * events on new prebuilds and updates of the Prebuilds page
 * storing of prebuild infos on prebuild creation instead of computing over and over again

- [ ] /werft with-clean-slate-deployment

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **alextugarev** after the PR has been reviewed.

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**
- **[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by jankeromnes** [comment]
Removing the automatic review request because this is still Draft. 😊

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **alextugarev** after the PR has been reviewed.

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**
- **[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by AlexTugarev** [comment]
/hold

... as this is based on another PR

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **alextugarev** after the PR has been reviewed.

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**
- **[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/server/ee/src/workspace/workspace-factory.ts
Comment (AlexTugarev) [inline components/server/ee/src/workspace/workspace-factory.ts:133]: TODO extract to the suggested service

File: components/server/src/workspace/workspace-starter.ts
Comment (AlexTugarev) [inline components/server/src/workspace/workspace-starter.ts:196]: TODO remove TODO

File: components/ws-manager-bridge/src/messagebus-integration.ts
Comment (AlexTugarev) [inline components/ws-manager-bridge/src/messagebus-integration.ts:44]: TODO change to LOCAL

**Comment by jankeromnes** [comment]
> Error: create: failed to create: secrets "sh.helm.release.v1.gitpod.v1" is forbidden: unable to create new content in namespace staging-at-prebuild-events because it is being terminated

I guess the old namespace needs to be cleaned up first.

**Review by gtsiolis — APPROVED** [review]
Seeing prebuilds pop-up on the _Prebuilds_ list feels great! UX looks good! 🌟 🔝

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 03a58e29a0ba7afe522c8bfc7fff93755a2a1fe1</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/5116#pullrequestreview-738053508" title="Approved">gtsiolis</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **alextugarev** after the PR has been reviewed.

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**
- **[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Review by jankeromnes — APPROVED** [review]
Many thanks @AlexTugarev! 👍

The UX also looks good to me (the Prebuilds page is so fast now! ⚡⚡ thank you so much for that)

I've left a few questions & suggestions in-line. Most notably, I think the events on "start prebuild" and on "start workspace"(!) should be triggered asynchronously, so as to:
- Not make these important API calls slower with extra `await`s
- Not make these important API calls fail if triggering an event throws

/hold

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/5116#pullrequestreview-738053508" title="Approved">gtsiolis</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/5116#pullrequestreview-738088302" title="Approved">jankeromnes</a>*

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]
- ~~[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/dashboard/src/projects/Prebuilds.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/Prebuilds.tsx:50]: I prefer IIFE in this case, which also allows adding breakpoints for debugging.

File: components/dashboard/src/projects/Prebuild.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/Prebuild.tsx:52]: nice catch!

File: components/gitpod-db/src/typeorm/entity/db-prebuild-info-entry.ts
Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/entity/db-prebuild-info-entry.ts:34]: https://github.com/gitpod-io/gitpod/blob/313fc79b91ac9fe9d17054beac363c8db43555d4/components/gitpod-db/src/typeorm/entity/db-workspace.ts#L39-L40 doesn't check for the structure at all, i.e. any JSON will be forwarded as `WorkspaceContext`. I prefer some more robustness.

File: components/server/ee/src/workspace/workspace-factory.ts
Comment (AlexTugarev) [reply components/server/ee/src/workspace/workspace-factory.ts:168]: 1. you are right, let's decouple it.
2. solved by 1.)

File: components/server/src/projects/projects-service.ts
Comment (AlexTugarev) [reply components/server/src/projects/projects-service.ts:163]: nice catch. don't recall touching it. must be an issue with resolving merge conflicts on rebase.

File: components/server/src/projects/projects-service.ts
Comment (AlexTugarev) [reply components/server/src/projects/projects-service.ts:146]: 👍🏻  for tracing. I

File: components/server/src/workspace/workspace-starter.ts
Comment (AlexTugarev) [reply components/server/src/workspace/workspace-starter.ts:207]: let's decouple it. 'don't see a reason to await on it.

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

File: components/dashboard/src/projects/Prebuild.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/Prebuild.tsx:52]: @jankeromnes, I've add a check to prevent it from hitting the db layer: https://github.com/gitpod-io/gitpod/pull/5116/files#diff-9c81815b0337b4e00a2ab1615cab6010a2497412f99dfc63c73a17f8bf28a90cR877

Can you invite me to that team please? I'd like to reproduce/verify the fix

Comment (jankeromnes) [reply components/dashboard/src/projects/Prebuild.tsx:52]: It's on my user 😬 but please feel free to move it to a team by hacking the DB

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/5116#issuecomment-909996829" title="Approved">AlexTugarev</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/5116#pullrequestreview-738053508" title="Approved">gtsiolis</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/5116#pullrequestreview-738088302" title="Approved">jankeromnes</a>*

Associated issue requirement bypassed by: *<a href="https://github.com/gitpod-io/gitpod/pull/5116#issuecomment-909996829" title="Approved">AlexTugarev</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [AlexTugarev,jankeromnes]
- ~~[components/ee/db-sync/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/db-sync/OWNERS)~~ [AlexTugarev,jankeromnes]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [AlexTugarev,jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [AlexTugarev,jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [AlexTugarev,jankeromnes]
- ~~[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)~~ [AlexTugarev,jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 1b9a810a8e32404584b4553fa3b110b91735739d</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/5116#issuecomment-909996829" title="Approved">AlexTugarev</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/5116#pullrequestreview-738053508" title="Approved">gtsiolis</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/5116#pullrequestreview-738088302" title="Approved">jankeromnes</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/5116#issuecomment-910013695" title="LGTM">svenefftinge</a>*

Associated issue requirement bypassed by: *<a href="https://github.com/gitpod-io/gitpod/pull/5116#issuecomment-909996829" title="Approved">AlexTugarev</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [AlexTugarev,jankeromnes,svenefftinge]
- ~~[components/ee/db-sync/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/db-sync/OWNERS)~~ [AlexTugarev,jankeromnes,svenefftinge]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [AlexTugarev,jankeromnes,svenefftinge]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [AlexTugarev,jankeromnes,svenefftinge]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [AlexTugarev,jankeromnes,svenefftinge]
- ~~[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)~~ [AlexTugarev,jankeromnes,svenefftinge]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by AlexTugarev** [comment]
Thanks @svenefftinge!

**Pull request merged**
